app/logical/source/extractor/inkbunny.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+# @see Source::URL::Inkbunny
+# @see https://wiki.inkbunny.net/wiki/API
+class Source::Extractor::Inkbunny < Source::Extractor
+  def self.enabled?
+    Danbooru.config.inkbunny_session.present?
+  end
+
+  def match?
+    Source::URL::Inkbunny === parsed_url
+  end
+
+  def image_urls
+    if parsed_url.image_url?
+      [parsed_url.to_s]
+    else
+      submission[:files].to_a.pluck(:file_url_full)
+    end
+  end
+
+  def page_url
+    parsed_url.page_url || parsed_referer&.page_url
+  end
+
+  def artist_name
+    submission[:username]
+  end
+
+  def profile_url
+    "https://inkbunny.net/#{submission[:username]}" if submission.present?
+  end
+
+  def user_url
+    "https://inkbunny.net/user.php?user_id=#{submission[:user_id]}" if submission.present?
+  end
+
+  def profile_urls
+    [profile_url, user_url].compact
+  end
+
+  def artist_commentary_title
+    submission[:title]
+  end
+
+  def artist_commentary_desc
+    submission[:description_bbcode_parsed]
+  end
+
+  def dtext_artist_commentary_desc
+    DText.from_html(artist_commentary_desc) do |element|
+      if element.name == "table"
+        mention = element.at ".widget_userNameSmall"
+        if mention.present?
+          element.name = "a"
+          element[:href] = mention[:href]
+          element.content = mention.content
+        end
+      end
+    end&.strip
+  end
+
+  def tags
+    submission[:keywords].to_a.map do |tag|
+      [tag[:keyword_name], "https://inkbunny.net/search_process.php?keyword_id=#{tag[:keyword_id]}"]
+    end
+  end
+
+  def submission_id
+    parsed_url.submission_id || parsed_referer&.submission_id
+  end
+
+  def submission
+    api_response[:submissions].to_a.first || {}
+  end
+
+  memoize def api_response
+    return {} unless submission_id.present?
+
+    params = {
+      sid: Danbooru.config.inkbunny_session,
```

Comment:
I'm slightly unsure about this part tbh.
[Docs](https://wiki.inkbunny.net/wiki/API#Login) say that `sid` should remain valid for several days after its last use, which probably would be good enough.
But, the correct way to do that is to login with username/password and cache the sid until it expires - which I don't really know how to implement here.

Reply:
Nijie is an example of how to do this. But it's better to avoid logging in unless we have to. Sites often have captchas or rate limits on their login page to prevent automation. The Nijie tests break half the time they try to log in over and over again and that triggers rate limiting.

Reply:
I changed it do username/password login. Now that I actually look at their API, I guess this is what they want us to do and it should be fine if we have to login repeatedly.